google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java

```diff
@@ -1057,6 +1058,71 @@ public Query endAt(@Nonnull DocumentSnapshot snapshot) {
 
   /** Build the final Firestore query. */
   StructuredQuery.Builder buildQuery() {
+    StructuredQuery.Builder structuredQuery = buildWithoutClientTranslation();
+    if (options.getLimitType().equals(LimitType.Last)) {
+      // Apply client translation for limitToLast.
+
+      if (!options.getFieldOrders().isEmpty()) {
+        structuredQuery.clearOrderBy();
+        for (FieldOrder order : options.getFieldOrders()) {
+          // Flip the orderBy directions since we want the last results
+          order =
+              new FieldOrder(
+                  order.fieldReference,
+                  order.direction.equals(Direction.ASCENDING)
+                      ? Direction.DESCENDING
+                      : Direction.ASCENDING);
+          structuredQuery.addOrderBy(order.toProto());
+        }
+      }
+
+      if (options.getStartCursor() != null) {
+        structuredQuery.clearEndAt();
+        // Swap the cursors to match the flipped query ordering.
+        Cursor cursor =
+            options
+                .getStartCursor()
+                .toBuilder()
+                .setBefore(!options.getStartCursor().getBefore())
+                .build();
+        structuredQuery.setEndAt(cursor);
+      }
+
+      if (options.getEndCursor() != null) {
+        structuredQuery.clearStartAt();
+        // Swap the cursors to match the flipped query ordering.
+        Cursor cursor =
+            options
+                .getEndCursor()
+                .toBuilder()
+                .setBefore(!options.getEndCursor().getBefore())
+                .build();
+        structuredQuery.setStartAt(cursor);
+      }
+    }
+
+    return structuredQuery;
+  }
+
+  /**
+   * Builds a query to saved in a bundle file.
+   *
+   * <p>This will not do client translation (limitToLast order flip, for example),
```

Comment:
```suggestion
   * <p>This will not perform any limitToLast order flip, as {@link BundledQuery} has first class
   * representation via {@link BundledQuery.LimitType}
```

Reply:
Done.